robot/lib/Src/drivers/ICM20948.cpp

```diff
@@ -0,0 +1,193 @@
+#include "drivers/ICM20948.hpp"
+#include <cmath>
+
+#define READ_BIT (1 << 7)
+
+#define WHOAMI_VAL 0xEA
+
+namespace Registers {
+    // Bank select
+    constexpr uint8_t REG_BANK_SEL  = 0x7F;
+
+    // Bank 0
+    constexpr uint8_t WHO_AM_I      = 0x00;
+    constexpr uint8_t USER_CTRL     = 0x03;
+    constexpr uint8_t LP_CONFIG     = 0x05; 
+    constexpr uint8_t PWR_MGMT_1    = 0x06;
+    constexpr uint8_t PWR_MGMT_2    = 0x07;
+
+    constexpr uint8_t GYRO_Z_OUT_H  = 0x37;
+    constexpr uint8_t GYRO_Z_OUT_L  = 0x38;
+
+    constexpr uint8_t FIFO_MODE     = 0x69;
+    //constexpr uint8_t 
+   
+    // Bank 2 
+    constexpr uint8_t GYRO_CONFIG_1 = 0x01;
+    //constexpr uint8_t 
+
+
+    enum GyroFS {
+        DPS_2000 = 3,
+        DPS_1000 = 2,
+        DPS_500 = 1,
+        DPS_250 = 0,
+    };
+
+};
+
+ICM20948::ICM20948(std::shared_ptr<SPI> imuSPI, PinName cs_pin) 
+    : imuSPI(std::move(imuSPI)),
+      nCs(cs_pin) {
+    //chip_select_pin.write(false);
+    nCs = 1;
+    HAL_Delay(1500);
+}
+
+bool ICM20948::initialize() {
+    // chip_select(true);
+
+    // SPI pulise to clear pins. 
+    imuSPI->transmit(0);    
+    
+    HAL_Delay(100);
+
+    uint8_t whoami = read_register(0, Registers::WHO_AM_I);
+    bool failed_init = false;
+
+    if (whoami != WHOAMI_VAL) {
+        printf("Failed to connect to IMU.");
+        failed_init = true;
+    }
+
+    // Reset
+    write_register(0, Registers::PWR_MGMT_1, 0xC1);
+
+    HAL_Delay(100);
+
+    // Check whoami again
+    whoami = read_register(0, Registers::WHO_AM_I);
+
+    if (whoami != WHOAMI_VAL && failed_init) {
```

Comment:
This should be `||`, right? Otherwise this condition will be always false (because `failed_init` is only `true` whenever the above condition has already returned from the function).

Also, this should _definitely_ retry until it works properly. The `HAL_Delay` isn't a very good way of timing the gyro's boot time.